src/system_identity/l1_token_retriever.rs

```diff
@@ -0,0 +1,127 @@
+use std::{fmt, time::Duration};
+
+use chrono::{TimeDelta, Utc};
+use http::{header::CONTENT_TYPE, StatusCode, Uri};
+use serde::Deserialize;
+use serde_json::json;
+
+use crate::{
+    http_client::{HttpClient, HttpClientError},
+    token::{Token, TokenType},
+    TokenRetriever, TokenRetrieverError,
+};
+
+pub struct L1TokenRetriever<'a, C: HttpClient> {
+    client_id: String,
+    client_secret: String,
+    http_client: &'a C,
+    token_retrieval_uri: &'a Uri,
+}
+
+impl<C: HttpClient> fmt::Debug for L1TokenRetriever<'_, C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("L1TokenRetriever")
+            .field("client_id", &self.client_id)
+            .field("client_secret", &self.client_secret)
+            .field("token_retrieval_uri", &self.token_retrieval_uri)
+            .field("http_client", &"impl HttpClient")
+            .finish()
+    }
+}
+
+impl<'a, C: HttpClient> L1TokenRetriever<'a, C> {
+    pub fn new(
+        client_id: String,
+        client_secret: String,
+        http_client: &'a C,
+        token_retrieval_uri: &'a Uri,
+    ) -> Self {
+        Self {
+            client_id,
+            client_secret,
+            http_client,
+            token_retrieval_uri,
+        }
+    }
+}
+
+impl<C: HttpClient> TokenRetriever for L1TokenRetriever<'_, C> {
+    fn retrieve(&self) -> Result<Token, TokenRetrieverError> {
+        token_from_client_secret(
+            self.http_client,
+            &self.client_id,
+            &self.client_secret,
+            self.token_retrieval_uri,
+        )
+        .map_err(|e| TokenRetrieverError::TokenRetrieverError(e.to_string()))
+    }
+}
+
+/// Basic response coming from the NR token retrieval endpoint
+#[derive(Debug, PartialEq, Deserialize)]
+struct TokenRetrievalResponse {
+    access_token: String,
+    expires_in: u64,
+    token_type: String,
+}
+
+impl TryFrom<TokenRetrievalResponse> for Token {
+    type Error = TokenRetrieverError;
+
+    fn try_from(response: TokenRetrievalResponse) -> Result<Self, Self::Error> {
+        let access_token = response.access_token;
+        let token_type = TokenType::try_from(response.token_type.as_str())
+            .map_err(|e| TokenRetrieverError::TokenRetrieverError(e.to_string()))?;
+
+        // Assuming we get seconds from the `expires_in` field of the JSON response
+        let time_delta = TimeDelta::from_std(Duration::from_secs(response.expires_in))
+            .map_err(|e| TokenRetrieverError::TokenRetrieverError(e.to_string()))?;
+
+        let expires_at = Utc::now().checked_add_signed(time_delta).ok_or_else(|| {
+            TokenRetrieverError::TokenRetrieverError(
+                "Failed to calculate expiration time".to_string(),
+            )
+        })?;
+
+        Ok(Token::new(access_token, token_type, expires_at))
+    }
+}
+
+fn token_from_client_secret<C: HttpClient>(
```

Comment:
Would it make sense to make it a method of `L1TokenRetriever`?

Reply:
I think I'm not a fan of methods if I am able to extract functionality to a (private to the module) free function so when looking at the signature I know exactly what I need and what I get back. But I'm not really consistent with that opinion (I have a static method somewhere that you pointed out at already, so there's room for discussion 😄)

Reply:
I'm normally opt for the same approach of static functions but if all the fields of the struct are received as arguments I think the function is actually a private method

Reply:
Hmm good catch, perhaps before I had more fields and ended up having actually the same so there's no meaningful distinction 🤔 I'll review that!

Reply:
Related with a previous comment, `retrieve` does nothing but call this static function. Can the code be inside retrieve?
```rust
    fn retrieve(&self) -> Result<Token, TokenRetrieverError> {
        token_from_client_secret(
            self.http_client,
            &self.client_id,
            &self.client_secret,
            self.token_retrieval_uri,
        )
        .map_err(|e| TokenRetrieverError::TokenRetrieverError(e.to_string()))
    }
```

Reply:
Changed in https://github.com/newrelic/newrelic-auth-rs/pull/25/commits/3f58cb9e612c9eeb64dab622dd78586327b78371, but this will change further on next commits.

Reply:
Probably I'm missing something, but still `retrieve` it's a wrapper for `retrieve_token`
```rust
fn retrieve(&self) -> Result<Token, TokenRetrieverError> {
  self.retrieve_token()
}
```
Is there any reason not to have the body of `retrieve_token` inside `retrieve` and just have one method? 🤔